File: Hard/140. Word Break II.cpp
class Solution {
public:
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        // Convert wordDict to a set for O(1) lookup
        unordered_set<string> wordSet(wordDict.begin(), wordDict.end());
        vector<string> results;
        string currentSentence;
        // Start the backtracking
        backtrack(s, wordSet, currentSentence, results, 0);
        return results;
    }



private:
    void backtrack(const string& s, const unordered_set<string>&wordSet, string& currentSentence, vector<string>& results, int startIndex) {
        // Base case: we have reached the end of the string
        // Add the current string to the result
        if (startIndex == s.length()) {
            results.push_back(currentSentence);
            return;
        }

        // Iterate over possible end indices
        for (int endIndex = startIndex + 1; endIndex <= s.length(); ++endIndex) {
            string word = s.substr(startIndex, endIndex - startIndex);
            // If the word is in the substring, continue with the backtrack
            if (wordSet.find(word) != wordSet.end()) {
                string originalSentence = currentSentence;
                if (!currentSentence.empty()) currentSentence += " ";
                currentSentence += word;
                // Recursively call backtrack with the new end index
                backtrack(s, wordSet, currentSentence, results, endIndex);
                
                // Reset currentSentence to it's original state (value)
                currentSentence = originalSentence;
            }
        }
    };
};

File: Hard/140. Word Break II.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        wordDict = set(wordDict)
        cache = {}


        def backTrack(i):
            if i == len(s):
                return [""]
            
            result = []

            for j in range(len(s)):
                word = s[i: j + 1]
                if word not in wordDict:
                    continue

                strings = backTrack(j + 1)

                if not strings:
                    continue

                for sub_string in strings:
                    sentence = word 
                    if sub_string:
                        sentence += " " + sub_string
                    result.append(sentence)
                
            cache[i] = result
            return result
        
        return backTrack(0)